Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp

```
@@ -1625,4 +1640,16 @@ void WebResourceLoadStatisticsStore::setStorageAccessPermissionForTesting(bool g
     });
 }
 
+void WebResourceLoadStatisticsStore::wasGrantedStorageAccessPermissionInPage(WebPageProxyIdentifier webPageProxyID, const RegistrableDomain& topFrameDomain, const RegistrableDomain& subFrameDomain)
+{
+    m_domainsGrantedStorageAccessPermissionInPage.ensure(webPageProxyID, [] {
+        return HashSet<std::pair<TopFrameDomain, SubFrameDomain>>();
+    }).iterator->value.add({ topFrameDomain, subFrameDomain });
+}
+
+void WebResourceLoadStatisticsStore::wasRevokedStorageAccessPermissionInPage(WebPageProxyIdentifier webPageProxyID)
```

For these kind of non-hot code paths, I like to submit the expected domains and check that the proxy ID matches the expected domains before it's deleted. Is that reasonable here or do you have cases where you're deleting access without knowledge of domains? Maybe you at least know the top frame domain in all cases? It's just good to have that and get loud bug reports if we mess it up rather than silent failures.

---

The case where this is called is when the tab associated with the WebPageProxy is closed, I don't think it makes sense to check the domains in that scenario.

---

I see two cases. We know the top frame and sub frame in one case, but we only have a WebPageProxyID in the other case. We could add an optional pair of RegistrableDomains and then only perform the check that John describes when we know them. Right now this is only used when running tests, but you could use this when deleting website data, too, like John suggested.

---

For deleting website data we don't have a WebPageProxyIdentifier though, so we couldn't reuse it for that. And yes, the other case is just used for testing.